File: pt-1/3871.count-commas-in-range-ii.java
/*
 * @lc app=leetcode id=3871 lang=java
 *
 * [3871] Count Commas in Range II
 *
 * Numbers from 1,000 to 999,999 have one comma, the next three-digit block has
 * two, and so on. Sum whole thousand-based ranges, clipping the final range at n.
 *
 * Time: O(log_1000 n). Space: O(1).
 */

// @lc code=start
class Solution {
    public long countCommas(long n) {
        long answer = 0;
        long start = 1000;
        long commas = 1;

        while (start <= n) {
            long end = start * 1000 - 1;
            long actualEnd = Math.min(n, end);
            answer += (actualEnd - start + 1) * commas;
            start *= 1000;
            commas++;
        }
        return answer;
    }
}
// @lc code=end

File: pt-1/3871.count-commas-in-range-ii.go
package leetcode

//
// @lc app=leetcode id=3871 lang=golang
//
// [3871] Count Commas in Range II
//
// Notes
// Numbers from 1,000 to 999,999 each contribute one comma, the next thousand
// block contributes two, and so on. Scan powers of 1000 and add the block size
// times its comma count. Time: O(log_1000 n). Space: O(1).
//
// @lc code=start

func CountCommas3871(n int) int {
	answer := 0
	start := 1000
	commas := 1

	for start <= n {
		end := start*1000 - 1
		actualEnd := end
		if actualEnd > n {
			actualEnd = n
		}
		answer += (actualEnd - start + 1) * commas
		start *= 1000
		commas++
	}

	return answer
}

// @lc code=end
